File: src/Enterprise/Kotlin/Example3.kt
// Abstract Product
interface Report {
    fun generate()
}

// Concrete Product
class PDFReport : Report {
    override fun generate() {
        println("Generating PDF report")
    }
}

class HTMLReport : Report {
    override fun generate() {
        println("Generating HTML report")
    }
}

// Abstract Factory
interface ReportFactory {
    fun createReport(): Report
}

// Concrete Factory
class PDFReportFactory : ReportFactory {
    override fun createReport(): Report {
        return PDFReport()
    }
}

class HTMLReportFactory : ReportFactory {
    override fun createReport(): Report {
        return HTMLReport()
    }
}

fun useFactory(factory: ReportFactory) {
    val report = factory.createReport()
    report.generate()
}

// Usage
fun main() {
    useFactory(PDFReportFactory())
    useFactory(HTMLReportFactory())
}

File: src/Scripting/RubyRB/example3.rb
# Abstract Factory
class ReportFactory
    def create_report
      raise NotImplementedError, 'create_report() must be implemented'
    end
  end
  
  # Concrete Factories
  class PDFReportFactory < ReportFactory
    def create_report
      PDFReport.new
    end
  end
  
  class HTMLReportFactory < ReportFactory
    def create_report
      HTMLReport.new
    end
  end
  
  # Products
  class PDFReport
    def generate
      puts 'Generating PDF report'
    end
  end
  
  class HTMLReport
    def generate
      puts 'Generating HTML report'
    end
  end
  
  def use_factory(factory)
    report = factory.create_report
    report.generate
  end
  
  # Usage
  use_factory(PDFReportFactory.new)
  use_factory(HTMLReportFactory.new)
  